### PR TITLE
Allow `context_name` var in `enabled_if` field of libraries

### PR DESCRIPTION
**Pull request by voodoos**
This answers a request from @samoht relative to Mirage dune's build.

However I am still bothered by the fact that the variables used in libs' `enabled_if` fields are not expanded using the standard mechanism (using and `Expander`), instead these vars have custom callback functions defined in `Lib_info`:

```ocaml
let var_map =
  [ ("architecture", fun t -> t.architecture)
  ; ("system", fun t -> t.system)
  ; ("model", fun t -> t.model)
  ; ("os_type", fun t -> Ocaml_config.Os_type.to_string t.os_type)
  ; ("ccomp_type", fun t -> Ocaml_config.Ccomp_type.to_string t.ccomp_type)
  ; ("profile", fun t -> Profile.to_string t.profile)
  ; ("ocaml_version", fun t -> t.ocaml_version_string)
  ; ("context_name", fun t -> Context_name.to_string t.context_name)
  ]
```

@rgrinberg, @jeremiedimino: Maybe you know the reason for that ? 
I guess one issue is that these variables are expanded during the creation of a context when `Dune_file.Library.to_lib_info` is called, and expanding vars normally requires a pre-existing `Context.t`. But what is the reason for expanding them immediately ?

### CODE REVIEW

**Comment by samoht** [comment]
Do you know if it's possible to define the same library in two different context (but using different dependencies or flags)?

**Comment by rgrinberg** [comment]
> Thanks! With this patch do you know if it's possible to define the same library in two different contexts (but using different dependencies or flags)?

It's possible to customize flags. Try using `(:include ..)` and then varying the flags based on the context_name in your configurator program.

File: test/blackbox-tests/test-cases/enabled_if/eif-forbidden_var.t/run.t [hunk 1]
```diff
@@ -0,0 +1,10 @@
+This one unse forbidden variables
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/enabled_if/eif-forbidden_var.t/run.t:1]: `unse` is a typo

**Review by rgrinberg — APPROVED** [review]
> instead these vars have custom callback functions defined in Lib_info:

You mean `Lib_config`?

> Maybe you know the reason for that ?

In the past `Expander` never existed and its functionality lived in a single massive function in `Super_context`(🤣 ). I believe at that point David or me found it easier to just add a primitive evaluation mechanism for enabled_if on the side.

> I guess one issue is that these variables are expanded during the creation of a context when Dune_file.Library.to_lib_info is called, and expanding vars normally requires a pre-existing Context.t. But what is the reason for expanding them immediately ?

There is no reason. We could easily have a `to_lib_info` take a parameter for the evaluated `enabled_if`. You should be able to use the toplevel `Expander` which forbids all dynamic forms as well.

**Comment by voodoos** [comment]
I am going to merge and will try to allocate time for the refactoring of this part soon.

**Pull request merged**